{"cells":[{"cell_type":"markdown","metadata":{"id":"DQg2BdFIg7TM"},"source":["# W200 Introduction to Data Science Programming, UC Berkeley MIDS\n","\n","\n","## Instructions\n","The final exam is designed to evaluate your grasp of Python theory as well as Python coding.\n","\n","- This is an individual exam.\n","- You have 48 hours to complete the exam, starting from the point at which you first access it.\n","- You will be graded on the quality of your answers.  Use clear, persuasive arguments based on concepts we covered in class.\n","- Please double-click the markdown cells where it says \"Your answer here\" to input answers (if you need more cells please make them markdown cells)\n","- Use only Python standard libraries, matplotlib, seaborn, NumPy and Pandas for this exam\n","- Upload your exam to gradescope and your github repo in the folder /SUBMISSIONS/final_exam/. **The code version uploaded to Part 2 will be your official version for grading purposes.**\n","- There is no autograder for this notebook - your exam will be manually graded.\n","- Please restart and run all cells before submitting so that all of your answers are shown. You do not need to upload the dataset into gradescope or github."]},{"cell_type":"markdown","metadata":{"id":"hfJDE2HWg7TT"},"source":["## YOUR NAME HERE"]},{"cell_type":"markdown","metadata":{"id":"Dxxzo-O4g7TU"},"source":["## 1: Short Answer Questions (25 pts - each question = 5 pts)\n","\n","a) The following class ```Cart``` and method ```add_to_cart``` are parts of a larger program used by a mobile phone company.  The method ```add_to_cart``` will work when an object of type MobileDevice or of type ServiceContract is passed to it. State whether the method ```add_to_cart``` is a demonstration of the following items (yes/no) and the reasoning (1-2 sentences): \n","\n","    1. Inheritance\n","    2. Polymorphism\n","    3. Object-Oriented programming\n","    4. Top-down design\n","    5. Functional programming"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"okLhQLwfg7TV"},"outputs":[],"source":["# Method:\n","class Cart():\n","    \n","    def __init__(self):\n","        self.cart = []\n","        self.total = 0\n","    \n","    def add_to_cart(self, item):\n","        self.cart.append(item)\n","        self.total += item.price"]},{"cell_type":"markdown","metadata":{"id":"syKK8Dj_g7Ta"},"source":["- a) Your answer here\n","\n","\n","1. Inheritance:\n","2. Polymorphism:\n","3. Object-Oriented programming:\n","4. Top-down design:\n","5. Functional programming:"]},{"cell_type":"markdown","metadata":{"id":"JdTWk7xZg7Ta"},"source":["b) Suppose you have a long list of digits (0-9) that you want to write to a file.  From a storage standpoint, would it be more efficient to use ASCII or UTF-8 as an encoding? What is the most efficient way to create an even smaller file to store the information?"]},{"cell_type":"markdown","metadata":{"id":"UDRs-eIxg7Tb"},"source":["- b) Your answer here"]},{"cell_type":"markdown","metadata":{"id":"RMIrgGNjg7Tc"},"source":["c) How would you sanity-check your data? Why is it important to sanity-check your data before you begin your analysis? What could happen if you don't?"]},{"cell_type":"markdown","metadata":{"collapsed":true,"id":"errKkCYQg7Td"},"source":["- c) Your answer here"]},{"cell_type":"markdown","metadata":{"id":"pY3aAxV-g7Te"},"source":["d) How do you determine which variables in your dataset you should check for issues prior to starting an analysis? "]},{"cell_type":"markdown","metadata":{"collapsed":true,"id":"doSOqeqog7Te"},"source":["- d) Your answer here"]},{"cell_type":"markdown","metadata":{"id":"GQjhe5avg7Tf"},"source":["e1) Explain why the following code prints what it does."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"j2cQqWirg7Tg"},"outputs":[],"source":["def g(): pass\n","print(type(g))"]},{"cell_type":"markdown","metadata":{"id":"YOrDq07ag7Th"},"source":["- e1) Your answer here"]},{"cell_type":"markdown","metadata":{"id":"MAs1nQZ_g7Th"},"source":["e2) Explain why the following code prints something different."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"1b49c_4Ag7Ti"},"outputs":[],"source":["def g(): pass\n","print(type(g()))"]},{"cell_type":"markdown","metadata":{"id":"9jBjkguyg7Tj"},"source":["- e2) Your answer here"]},{"cell_type":"markdown","metadata":{"id":"jtHAe_5Pg7Tj"},"source":["## 2:  General Coding Questions (15 pts - each question 5 pts)"]},{"cell_type":"markdown","metadata":{"id":"2EAQnoKwg7Tj"},"source":["a) Using a dictionary comprehension, make a dictionary with the keys being the lowercase vowels (aeiou) and the values being the number of vowels in the example_text (case insensitive). Fill in a dictionary comprehension below so that we get this desired output.\n","\n","```\n","example_text = 'Outlook is bright!'\n","```\n","\n","expected output:\n","```\n","{'i': 2, 'o': 3, 'u': 1}\n","\n","or\n","\n","{'a': 0, 'e': 0, 'i': 2, 'o': 3, 'u': 1}\n","```"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"yT_f_hhgg7Tk"},"outputs":[],"source":["example_text = \"I love coding in python!\"\n","\n","\n","# 2a) Your code here"]},{"cell_type":"markdown","metadata":{"id":"j5v3mwmVg7Tk"},"source":["b) Below is a data frame of customers that have different cooling systems. Your data science team lead wants the column ```cooling_system``` to be labeled with the integer numbers 1-4 instead of the text as shown below:\n","\n","- 1 = Air Conditioning / Air Con / AC\n","- 2 = Heat Pump / HP\n","- 3 = Evaporative Cooler / EvapCooler / EC\n","- 4 = Fan\n","\n","Make a new column called ```cooling_type``` that maps the text values to the new numeric values. Filter out / remove rows with the values that are not included in the mapping above. Print out/display this new data frame. Be sure to list any assumptions also!"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"61EFcCtbg7Tl"},"outputs":[],"source":["import pandas\n","\n","# creating a data frame from scratch - list of lists\n","\n","data = [ \n","         [101, 'AC'], \n","         [102, 'Heat Pump'], \n","         [103, 'Air Con'], \n","         [104, 'Air Conditioning'],\n","         [105, 'Fan'],\n","         [106, 'None'],\n","         [107, 'EvapCooler'],\n","         [108, None],\n","         [109, 'EC'],\n","         [110, 'Evaporative Cooler'],\n","         [111, 'geothermal'],\n","         [112, 8],\n","         [113, 'Air Con']\n","       ]\n","\n","# create a data frame with column names - list of lists\n","\n","col_names = ['Cust_Number', 'Cooling_System']\n","df = pandas.DataFrame(data, columns=col_names)\n","df"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"E3IH8MHwg7Tl"},"outputs":[],"source":["# 2b) Your code here"]},{"cell_type":"markdown","metadata":{"id":"LzguKp3Rg7Tm"},"source":["c) From the dataframe below, use groupby in Pandas to show how many total delegates were obtained grouped by state. Print this out."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"SBKSw6hsg7Tm"},"outputs":[],"source":["import pandas\n","\n","# creating a data frame from scratch - list of lists\n","\n","data = [ ['marco', 165, 'blue', 'FL'], \n","         ['jeb', 35, 'red', 'FL'], \n","         ['chris', 0, 'white', 'NY'], \n","         ['donald', 1543, 'white', 'NY'],\n","         ['ted', 559, 'blue', 'NY'],\n","         ['john', 161, 'red', 'OH']\n","       ]\n","\n","# create a data frame with column names - list of lists\n","\n","col_names = ['name', 'delegates', 'color', 'state']\n","df = pandas.DataFrame(data, columns=col_names)\n","df"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"KNNZmJGUg7To"},"outputs":[],"source":["# 2c) Your code here"]},{"cell_type":"markdown","metadata":{"id":"j0vXNb_Yg7Tp"},"source":["## 3: Bitcoin coding problem (20 points):"]},{"cell_type":"markdown","metadata":{"id":"PWU5dtSEg7Tp"},"source":["#### Bitcoin\n","\n","Consider a record of a one-time investment in bitcoin with value of that investment tracked monthly, provided as an (ordered) tuple of dictionaries, where each dictionary comprises one key for the month and corresponding one value for the value of the investment as of that date. (For example 31 Jan 2018 was $1200 as shown in `data` below).\n","\n","The initial investment is made on 1 Jan 2018 and was for $1000 (labelled 'Initial' in the `data` below).\n","\n","Fill in the Python function (named `bitcoin_invest` below) to take such a record of any length (the below `data` is only a sample but the full data will be in the same format), and return a dataframe comprising a row for each month with columns for date, balance, and return. **Print out/display this dataframe.**\n","\n","Also, visualize the record as two vertically arranged plots (there are a few ways to do this, keep it consistent over both plots). \n","- The top plot should show a line plot of balance vs. month \n","- The bottom plot should show a bar plot of return vs. month, with a black horizontal line at return=0, and bars color-coded such that positive returns are green and negative returns are red.  \n","- The two plots' horizontal axes should align. Demonstrate that your code works by applying it to `data`.\n","\n","Some definitions: <br>\n","- The gain for each period is the end balance minus the start balance. \n","- The growth factor for each period is the end balance divided by the start balance.  \n","- The return for each period is the growth factor minus 1."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"nNyReUHXg7Tp"},"outputs":[],"source":["data = ({\"Initial\": 1000}, {\"Jan 2018\":1200},{\"Feb 2018\":1400},{\"Mar 2018\":700},{\"Apr 2018\":800},{\"May 2018\":500})\n","data"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"tqPVdbgkg7Tp"},"outputs":[],"source":["# 3) Your code here\n","\n","def bitcoin_invest(monthly_data):\n","    # add code here"]},{"cell_type":"markdown","metadata":{"id":"qGnnkRF9g7Tq"},"source":["## 4: Clinical disease data (40 pts)\n","\n","Your boss comes to you Monday morning and says “I figured out our next step; we are going to pivot from an online craft store and become a data center for genetic disease information! I found **ClinVar** which is a repository that contains expert curated data, and it is free for the taking. This is a gold mine! Look at the file and tell me what gene and mutation combinations are classified as dangerous.”\n","\n","Make sure that you only give your boss the dangerous mutations and include:\n","\n","1) Gene name\n","\n","2) Mutation ID number\n","\n","3) Mutation Position (chromosome & position)\n","\n","4) Mutation value (reference & alternate bases)\n","\n","5) Clinical significance (CLNSIG)\n","\n","6) Disease that is implicated\n","\n","**Requirements**\n","\n","1) The deliverables are the final result as a dataframe with a short discussion of any specifics. (that is, what data you would present to your boss with the explanation of your results)\n","\n","2) Limit your output to the first 100 harmful mutations and tell your boss how many total harmful mutations were found in the file\n","\n","3) Use the instructor-modified \"clinvar_final.txt\" at this link: https://drive.google.com/file/d/1Zps0YssoJbZHrn6iLte2RDLlgruhAX1s/view?usp=sharing This file was modified to be not exactly the same as 'standard' .vcf file to test your data parsing skills. **This is a large file so do NOT upload it into your github repo!**\n","\n","4) Replace missing values in the dataframe with: 'Not_Given'. Print or display this (including the Not_Given count) for the column `CLNSIG` by using pandas value_counts() function (https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html).\n","\n","5) State in your answer how you define harmful mutations\n","\n","**6) Do your best on getting to above requirements and submit whatever you do before the deadline. If your work is incomplete be sure to describe the blockers that got in your way and how you might get past them (if given more time).**\n","\n","7) You can use as many code blocks as you need. Please clean-up your code and make it readable for the graders!\n","\n","**Hints** \n","* We do not expect you to have any medical knowledge to solve this problem; look at the data, read the documentation provided, and write down your assumptions!\n","\n","* Correct pseudocode will give you partial credit so start with that. \n","\n","* Map out which fields you want to extract: Are they in the same place every time? What strategy will you use to robustly extract and filter your data of interest? How do you plan to handle missing data?\n","\n","* A good way to start is to print out each line, then practice parsing them to see if you can recover the fields of interest\n","\n","* A starting solution for parsing .vcfs can be found here: https://gist.github.com/dceoy/99d976a2c01e7f0ba1c813778f9db744 This solution does **NOT** work due to the changes we've made but can be modified to work. As with any solution that needs modifications, it may take less time to make your own solution!\n","\n","* Filter out junk and lines with no mutation data. Just focus on the data your need to deliver to your boss. \n","\n","* Pandas and NumPy parsers correctly recognize the end of each line in in the ClinVar file.\n","\n","* The unit of observation of this dataset is one row per mutation.\n","\n","* This is similar to a task that one of us tackled at work. You can answer the question with the information provided below or using the (partial) data dictionary file at this link: https://drive.google.com/file/d/1lx9yHdlcqmU_OlHiTUXKC_LQDqYBypH_/view?usp=sharing. Our goal is to see that you can put together a sensible plan, implement a solid parsing strategy, and document and justify the decisions that you made."]},{"cell_type":"markdown","metadata":{"collapsed":true,"id":"vxoh6uLEg7Tr"},"source":["### VCF file description (Summarized from version 4.1)\n","\n","```\n","* The VCF specification:\n","\n","VCF is a text file format which contains meta-information lines, a header line, and then data lines each containing information about a position in the genome. The format also can contain genotype information on samples for each position.\n","\n","* Fixed fields:\n","\n","There are 8 fixed fields per record. All data lines are **tab-delimited**. In all cases, missing values are specified with a dot (‘.’). \n","\n","1. CHROM - chromosome number\n","2. POS - position DNA nuceleotide count (bases) along the chromosome\n","3. ID - The unique identifier for each mutation\n","4. REF - reference base(s)\n","5. ALT - alternate base(s)\n","6. FILTER - filter status\n","7. QUAL - quality\n","8. INFO - a semicolon-separated series of keys with values in the format: <key>=<data>\n","\n","```\n","### Applicable INFO field specifications\n","\n","```\n","GENEINFO = <Gene name>\n","CLNSIG =  <Clinical significance>\n","CLNDN = <Disease name>\n","```\n","\n","### Sample ClinVar data (vcf file format - not exactly the same as the file to download!)\n","\n","```\n","##fileformat=VCFv4.1\n","##fileDate=2019-03-19\n","##source=ClinVar\n","##reference=GRCh38\t\t\t\t\t\t\t\n","#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\n","1\t949523\trs786201005\tC\tT\t.\t.\tGENEINFO=ISG15;CLNSIG=5\n","1\t949696\trs672601345\tC\tCG\t.\t.\tGENEINFO=ISG15;CLNSIG=5;CLNDN=Cancer\n","1\t949739\trs672601312\tG\tT\t.\t.\tGENEINFO=ISG15;CLNDBN=Cancer\n","1\t955597\trs115173026\tG\tT\t.\t.\tGENEINFO=AGRN;CLNSIG=2; CLNDN=Cancer\n","1\t955619\trs201073369\tG\tC\t.\t.\tGENEINFO=AGG;CLNDN=Heart_dis \n","1\t957640\trs6657048\tC\tT\t.\t.\tGENEINFO=AGG;CLNSIG=3;CLNDN=Heart_dis \n","1\t976059\trs544749044\tC\tT\t.\t.\tGENEINFO=AGG;CLNSIG=0;CLNDN=Heart_dis \n","```"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"qPhDns3zg7Tr"},"outputs":[],"source":["# 4) Your code here - can use as many code blocks as you would like"]},{"cell_type":"markdown","metadata":{"id":"QvWAQKYag7Tr"},"source":["4) Please Write your assumptions here:"]},{"cell_type":"markdown","metadata":{"id":"_r9uXf94g7Ts"},"source":["4) Findings / What would you present to your boss?"]}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.8.11"},"colab":{"provenance":[]}},"nbformat":4,"nbformat_minor":0}